src/WsReconnect/index.ts

```diff
@@ -0,0 +1,97 @@
+import TimeNode from '../TimeNode';
+import { ReconnectMsg } from '../Enum';
+import W3Util from '../Util';
+
+declare const setTimeout: any;
+
+export default class WsReconnect {
+  private timenode: TimeNode;
+  private reconnectTries: number = 0;
+  private reconnecting: boolean = false;
+  private reconnected: boolean = false;
+
+  constructor(timenode: TimeNode) {
+    this.timenode = timenode;
+  }
+
+  public setup(): void {
+    const {
+      logger,
+      web3: { currentProvider }
+    } = this.timenode.config;
+
+    currentProvider.on('error', (err: any) => {
+      logger.debug(`[WS ERROR] ${err}`);
+      setTimeout(async () => {
+        const msg: ReconnectMsg = await this.handleWsDisconnect();
+        logger.debug(`[WS RECONNECT] ${msg}`);
+      }, this.reconnectTries * 1000);
+    });
+
+    currentProvider.on('end', (err: any) => {
+      logger.debug(`[WS END] Type= ${err.type} Reason= ${err.reason}`);
+      setTimeout(async () => {
+        const msg = await this.handleWsDisconnect();
+        logger.debug(`[WS RECONNECT] ${msg}`);
+      }, this.reconnectTries * 1000);
+    });
+  }
+
+  private async handleWsDisconnect(): Promise<ReconnectMsg> {
```

Comment:
I think we should pass in the error type as an argument,
to help for the test and better handling of the individual scenarios.

Reply:
That's not a bad idea. I think we're going to merge this now as a "good enough" solution but let's continue to work on this in a separate PR. @adibas03